File: src/CCVTAC.Console/ResultTracker.cs
namespace CCVTAC.Console;

/// <summary>
/// Tracks the successes and failures of various operations.
/// Successes are only counted; failure error messages are tracked.
/// </summary>
internal sealed class ResultTracker<T>
{
    private nuint _successCount;
    private readonly Dictionary<string, string> _failures = [];
    private readonly Printer _printer;

    private static string CombineErrors(Result<T> result) =>
        string.Join(" / ", result.Errors.Select(e => e.Message));

    internal ResultTracker(Printer printer)
    {
        ArgumentNullException.ThrowIfNull(printer);
        _printer = printer;
    }

    /// <summary>
    /// Logs the result for a specific corresponding input.
    /// </summary>
    internal void RegisterResult(string input, Result<T> result)
    {
        if (result.IsSuccess)
        {
            _successCount++;
            return;
        }

        string errors = CombineErrors(result);
        if (!_failures.TryAdd(input, errors))
        {
            // Keep only the latest error for each input.
            _failures[input] = errors;
        }
    }

    /// <summary>
    /// Prints any failures for the current batch.
    /// </summary>
    internal void PrintBatchFailures()
    {
        if (_failures.Count == 0)
        {
            _printer.Debug("No failures in batch.");
            return;
        }

        var failureLabel = _failures.Count == 1 ? "failure" : "failures";
        _printer.Info($"{_failures.Count} {failureLabel} in this batch:");

        foreach (var (url, error) in _failures)
        {
            _printer.Warning($"- {url}: {error}");
        }
    }

    /// <summary>
    /// Prints the output for the current application session.
    /// Expected to be used upon quitting.
    /// </summary>
    internal void PrintSessionSummary()
    {
        var successLabel = _successCount == 1 ? "success" : "successes";
        var failureLabel = _failures.Count == 1 ? "failure" : "failures";

        _printer.Info(
            $"Quitting with {_successCount} {successLabel} and {_failures.Count} {failureLabel}."
        );

        foreach (var (url, error) in _failures)
        {
            _printer.Warning($"- {url}: {error}");
        }
    }
}

File: src/CCVTAC.FSharp/ResultTracker.fs
namespace CCVTAC.Console

open System
open System.Collections.Generic

type ResultTracker<'a>(printer: Printer) =

    let mutable successCount : uint64 = 0UL

    let failures = Dictionary<string,string>()

    static let combineErrors (errors: string list) =
        String.Join(" / ", errors)

    member _.AddSuccess() : unit =
        successCount <- successCount + 1UL

    member _.AddFailure(url: string, error: string) : unit =
        failures.Add(url, error)

    /// Logs the result for a specific corresponding input.
    member _.RegisterResult(input: string, result: Result<'a, string>) : unit =
        match result with
        | Ok _ ->
            successCount <- successCount + 1UL
        | Error e ->
            if not (failures.TryAdd(input, e)) then
                failures[input] <- e

    /// Prints any failures for the current batch.
    member _.PrintBatchFailures() : unit =
        if Numerics.isZero failures.Count then
            printer.Debug "No failures in batch."
        else
            let failureLabel = String.pluralize "failure" "failures" failures.Count
            printer.Info $"%d{failures.Count} %s{failureLabel} in this batch:"
            for pair in failures do
                printer.Warning $"- %s{pair.Key}: %s{pair.Value}"

    /// Prints the output for the current application session.
    member _.PrintSessionSummary() : unit =
        let successLabel = String.pluralize "success" "successes" successCount
        let failureLabel = String.pluralize "failure" "failures" failures.Count

        printer.Info $"Quitting with %d{successCount} %s{successLabel} and %d{failures.Count} %s{failureLabel}."

        for pair in failures do
            printer.Warning $"- %s{pair.Key}: %s{pair.Value}"
